### DISCUSSION TITLE
Segmentation fault (core dumped) error during the 3rd order Force constant calculation

### DISCUSSION THREAD

**Discussion by ballal09** [ttadano/alamode#47] [Q&A] [answered]
Dear Dr. Tadano,

I am doing a 3rd order force constant calculation for my structure which have 120 atoms in the supercell. The 3rd order force constant calculation went well  up to 3rd nearest neighbor cutoff distance but when I increased the cut off distance to 4th nearest neighbor and above (I am running a convergence test  with respect to cut off distance), the Alm code stops giving me the following error message:

**terminate called after throwing an instance of 'std::bad_alloc'**
**what():  std::bad_alloc**
**Aborted (core dumped)**

I know it happens due to memory issue. I have tried running it both in my PC and supercomputing cluster but no luck (same error persists). I have attached my  input file and last part of the logfile for your reference.  Please let me know your valuable suggestions. Thank you very much for your time.

Best
Ballal

Input file:

&general
  PREFIX = LYC114_cubic
  MODE = optimize
  NAT = 120; NKD = 3
  KD = Li Y Cl
/

&optimize
 DFSET = DFSET_cubic
 FC2XML = LYC114.xml
/

&interaction
  NORDER = 2  # 1: harmonic, 2: cubic, ..
/

&cell
  1.0 # factor in Bohr unit
  21.134085522436940607 -0.0 -0.0 # a1
 -10.56704276112398766 18.302654948131557688 -0.0 # a2
  -0.0 -0.0 46.401745214615175428 # a3
/

&cutoff 
  Li-Li None 9.17
  Li-Y  None 9.17
  Li-Cl None 5.102
  Y-Y   None 13.61
  Y-Cl  None 9.73
  Cl-Cl None 7.28
  
/

&position
1 0.0000000000000000 0.3429361544215348 0.0000000000000000
1 0.0000000000000000 0.3429361544215348 0.2500000000000000
1 0.0000000000000000 0.3429361544215348 0.5000000000000000
1 0.0000000000000000 0.3429361544215348 0.7500000000000000
1 0.6570638455784722 0.6570638455784722 -0.0000000000000000
1 0.6570638455784722 0.6570638455784722 0.2500000000000000
1 0.6570638455784722 0.6570638455784722 0.5000000000000000
1 0.6570638455784722 0.6570638455784722 0.7500000000000000
1 0.3429361544215348 0.0000000000000000 -0.0000000000000000
1 0.3429361544215348 0.0000000000000000 0.2500000000000000
1 0.3429361544215348 0.0000000000000000 0.5000000000000000
1 0.3429361544215348 0.0000000000000000 0.7500000000000000
1 0.0000000000000000 0.6612740425289326 -0.0000000000000000
1 0.0000000000000000 0.6612740425289326 0.2500000000000000
1 0.0000000000000000 0.6612740425289326 0.5000000000000000
1 0.0000000000000000 0.6612740425289326 0.7500000000000000
1 0.3387259874710628 0.3387259874710628 -0.0000000000000000
1 0.3387259874710628 0.3387259874710628 0.2500000000000000
1 0.3387259874710628 0.3387259874710628 0.5000000000000000
1 0.3387259874710628 0.3387259874710628 0.7500000000000000
1 0.6612740425289326 0.0000000000000000 -0.0000000000000000
1 0.6612740425289326 0.0000000000000000 0.2500000000000000
1 0.6612740425289326 0.0000000000000000 0.5000000000000000
1 0.6612740425289326 0.0000000000000000 0.7500000000000000
1 0.0000000000000000 0.6785788635649793 0.1250000000000000
1 0.0000000000000000 0.6785788635649793 0.3750000000000000
1 0.0000000000000000 0.6785788635649793 0.6250000000000000
1 0.0000000000000000 0.6785788635649793 0.8750000000000000
1 0.3214211664350302 0.3214211664350302 0.1250000000000000
1 0.3214211664350302 0.3214211664350302 0.3750000000000000
1 0.3214211664350302 0.3214211664350302 0.6250000000000000
1 0.3214211664350302 0.3214211664350302 0.8750000000000000
1 0.6785788635649793 0.0000000000000000 0.1250000000000000
1 0.6785788635649793 -0.0000000000000000 0.3750000000000000
1 0.6785788635649793 -0.0000000000000000 0.6250000000000000
1 0.6785788635649793 0.0000000000000000 0.8750000000000000
2 0.3333329860000021 0.6666669849999991 0.1271999316413147
2 0.3333329860000021 0.6666669849999991 0.3771999316413148
2 0.3333329860000021 0.6666669849999991 0.6271999316413147
2 0.3333329860000021 0.6666669849999991 0.8771999316413147
2 0.6666669849999991 0.3333329860000021 0.1228000763586859
2 0.6666669849999991 0.3333329860000021 0.3728000683586852
2 0.6666669849999991 0.3333329860000021 0.6228000683586853
2 0.6666669849999991 0.3333329860000021 0.8728000683586853
2 0.0000000000000000 -0.0000000000000000 -0.0000000000000000
2 0.0000000000000000 -0.0000000000000000 0.2500000000000000
2 0.0000000000000000 -0.0000000000000000 0.5000000000000000
2 0.0000000000000000 0.0000000000000000 0.7500000000000000
3 0.1067786918816598 0.5605779415444466 0.1886622725678447
3 0.1067786918816598 0.5605779415444466 0.4386622725678446
3 0.1067786918816598 0.5605779415444466 0.6886622725678448
3 0.1067786918816598 0.5605779415444466 0.9386622725678448
3 0.4394220584555533 0.5462007503371774 0.1886622725678447
3 0.4394220584555533 0.5462007503371774 0.4386622725678446
3 0.4394220584555533 0.5462007503371774 0.6886622725678448
3 0.4394220584555533 0.5462007503371774 0.9386622725678448
3 0.4537992496628226 0.8932213081183402 0.1886622725678447
3 0.4537992496628226 0.8932213081183402 0.4386622725678446
3 0.4537992496628226 0.8932213081183402 0.6886622725678448
3 0.4537992496628226 0.8932213081183402 0.9386622725678448
3 0.5605779415444466 0.1067786918816598 0.0613377314321557
3 0.5605779415444466 0.1067786918816598 0.3113377274321554
3 0.5605779415444466 0.1067786918816598 0.5613377274321552
3 0.5605779415444466 0.1067786918816598 0.8113377274321552
3 0.5462007503371774 0.4394220584555533 0.0613377314321557
3 0.5462007503371774 0.4394220584555533 0.3113377274321554
3 0.5462007503371774 0.4394220584555533 0.5613377274321552
3 0.5462007503371774 0.4394220584555533 0.8113377274321552
3 0.8932213081183402 0.4537992496628226 0.0613377314321557
3 0.8932213081183402 0.4537992496628226 0.3113377274321554
3 0.8932213081183402 0.4537992496628226 0.5613377274321552
3 0.8932213081183402 0.4537992496628226 0.8113377274321552
3 0.1171934047856111 0.2317844472320864 0.1932276147681892
3 0.1171934047856111 0.2317844472320864 0.4432276147681891
3 0.1171934047856111 0.2317844472320864 0.6932275847681870
3 0.1171934047856111 0.2317844472320864 0.9432275847681870
3 0.7682155827679160 0.8854089435535448 0.1932276147681892
3 0.7682155827679160 0.8854089435535448 0.4432276147681891
3 0.7682155827679160 0.8854089435535448 0.6932275847681870
3 0.7682155827679160 0.8854089435535448 0.9432275847681870
3 0.1145910344464533 0.8828065802143913 0.1932276147681892
3 0.1145910344464533 0.8828065802143913 0.4432276147681891
3 0.1145910344464533 0.8828065802143913 0.6932275847681870
3 0.1145910344464533 0.8828065802143913 0.9432275847681870
3 0.2317844472320864 0.1171934047856111 0.0567723892318111
3 0.2317844472320864 0.1171934047856111 0.3067723852318109
3 0.2317844472320864 0.1171934047856111 0.5567724152318130
3 0.2317844472320864 0.1171934047856111 0.8067724152318130
3 0.8854089435535448 0.7682155827679160 0.0567723892318111
3 0.8854089435535448 0.7682155827679160 0.3067723852318109
3 0.8854089435535448 0.7682155827679160 0.5567724152318130
3 0.8854089435535448 0.7682155827679160 0.8067724152318130
3 0.8828065802143913 0.1145910344464533 0.0567723892318111
3 0.8828065802143913 0.1145910344464533 0.3067723852318109
3 0.8828065802143913 0.1145910344464533 0.5567724152318130
3 0.8828065802143913 0.1145910344464533 0.8067724152318130
3 0.2115893656058012 0.7718956086565362 0.0667422980436697
3 0.2115893656058012 0.7718956086565362 0.3167422980436698
3 0.2115893656058012 0.7718956086565362 0.5667422980436696
3 0.2115893656058012 0.7718956086565362 0.8167422980436696
3 0.2281043913434639 0.4396937419492816 0.0667422980436697
3 0.2281043913434639 0.4396937419492816 0.3167422980436698
3 0.2281043913434639 0.4396937419492816 0.5667422980436696
3 0.2281043913434639 0.4396937419492816 0.8167422980436696
3 0.5603062580507113 0.7884106493941964 0.0667422980436697
3 0.5603062580507113 0.7884106493941964 0.3167422980436698
3 0.5603062580507113 0.7884106493941964 0.5667422980436696
3 0.5603062580507113 0.7884106493941964 0.8167422980436696
3 0.7718956086565362 0.2115893656058012 0.1832577019563303
3 0.7718956086565362 0.2115893656058012 0.4332577019563302
3 0.7718956086565362 0.2115893656058012 0.6832577019563304
3 0.7718956086565362 0.2115893656058012 0.9332577019563304
3 0.4396937419492816 0.2281043913434639 0.1832577019563303
3 0.4396937419492816 0.2281043913434639 0.4332577019563302
3 0.4396937419492816 0.2281043913434639 0.6832577019563304
3 0.4396937419492816 0.2281043913434639 0.9332577019563304
3 0.7884106493941964 0.5603062580507113 0.1832577019563303
3 0.7884106493941964 0.5603062580507113 0.4332577019563302
3 0.7884106493941964 0.5603062580507113 0.6832577019563304
3 0.7884106493941964 0.5603062580507113 0.9332577019563304

/

**The Alm code stops after the following section in the log file**

OPTIMIZATION
 ============

  LMODEL = least-squares

  Training data file (DFSET) : DFSET_cubic

  NSTART = 1; NEND = 2147
  2147 entries will be used for training.

  Total Number of Parameters : 22051
  Total Number of Free Parameters : 3353

  Entering fitting routine: SVD with constraints considered algebraically.

**Comment by ttadano** [comment, marked as answer]
To store the sensing matrix A, you will need, at least, 3353 x 120 x 3 x 2147 x 8 (Byte) = 20.7 GB of RAM. 

Here are suggested solutions:
- If you used the finite-difference approach for generating the displaced configurations, many elements of the displacements (in DFSET) are zero. In this case, you can reduce the RAM requirement by using a sparse solver (Please set ``SPARSE = 1`` in the &optimize field.)
- Use a computer node having more RAM. I believe many modern systems have RAM of, at least, 128 GB/node.
- Use compressive sensing option (``LMODEL = enet | adaptive-lasso``) in combination with DFSET generated for randomly sampled structures. This method enables us to reduce the number of  #displacement patterns dramatically, which also reduce the RAM requirement. I have not described details of this option in tutorial, but you can have a look at the materials at https://github.com/ttadano/alamode/tree/gh-pages/files.

**Reply by ballal09** [reply]
Dear Dr. Tadano

Thank you very much for your quick response, This was really helpful. Using SPARSE=1 , actually solved my problem. I have used the finite displacement approach for generating the structure.  I didn't find enough information  on how to generate randomly sampled structure in the file you mentioned but I am really interested to know that. 

Apart from that I have one more query, after calculating the cubic force constant when I am solving for thermal conductivity using the anphon program, I am getting the following warning: 
**OpenBLAS Warning : Detect OpenMP Loop and this application may hang. Please rebuild the library with USE_OPENMP=1 option**.

It does not stop my calculation. But it takes  long time (like 5 hr in my pc, although the configuration of my pc is not that really good) even for small K-grid (like 4 4 4).  I am using alamode inside the conda environment. While building the alamode using cmake , I have used the USE_OPENMP=1 option like **cmake -DUSE_MKL_FFT=no USE_OPENMP=1 ..**, but it didn't solve the warning.  Please let me know your valuable suggestion in your convenient time. I have attached my input file for the thermal conductivity calculation for your reference. Thank you very much for your time.

Best
Ballal Ahammed


Input file:


&general
  PREFIX = LYC114_cubic
  MODE   = RTA
  FCSXML = LYC114_cubic.xml
  NKD = 3; KD = Li Y Cl
  MASS = 6.94 88.906 35.45
  TMIN = 0; TMAX = 1000; DT = 200
  EMIN = 0; EMAX = 1000; DELTA_E = 10
/

&cell
  1.0 # factor in Bohr unit
  21.134085525899973135 -0.0 -0.0 # a1
 -10.567042762855503923 18.302654951130616467 -0.0 # a2
  -0.0 -0.0 46.401745225513394644 # a3
/
&kpoint
  2  # KPMODE = 1: line mode
  8 8 8
/

**Reply by ttadano** [reply]
Probably, you will have to rebuild OpenBlas with `USE_OPENMP=1`. If you have build alamode using an anaconda environment, which version of libopenblas are you using? (This can be checked by `conda list`).

**Reply by ballal09** [reply]
Dear Dr. Tadano

I see. I will give a try to rebuild OpenBlas with USE_OPENMP=1. FYI, the version of my  libopenblas is   0.2.20 . Thank you very much for your quick response.

Best
Ballal Ahammed

**Reply by chunweizhu** [reply]
Hi @ballal09
 `OpenBLAS Warning : Detect OpenMP Loop and this application may hang. Please rebuild the library with USE_OPENMP=1 option.` Have you solved this problem? I am also facing the same problem, can you give me some guidance? Thanks!
 the version of my libopenblas is 
 `libopenblas               0.3.23          pthreads_h80387f5_0    conda-forge`
Best

**Reply by ballal09** [reply]
Hi @chunweizhu 
I tried to rebuild  OpenBlas with USE_OPENMP=1, but it didn't work for me. I am sorry and I  hope you find some ways. Thank you.

**Reply by chunweizhu** [reply]
Hi @ballal09 , Regardless, thank you for your response.😊
